packages/pyright-internal/src/analyzer/dataClasses.ts

```diff
@@ -654,6 +673,93 @@ export function synthesizeDataClassMethods(
     );
 }
 
+// Validates converter and, if valid, returns its input type. If invalid,
+// fieldType is returned.
+function getConverterInputType(
+    evaluator: TypeEvaluator,
+    converterNode: ArgumentNode,
+    fieldType: Type,
+    fieldName: string
+): Type {
+    const converterType = evaluator.getTypeOfExpression(converterNode.valueExpression).type;
+
+    if (!isFunction(converterType) && !isOverloadedFunction(converterType)) {
+        return fieldType;
+    }
+
+    // Create synthesized function of the form Callable[[T], fieldType] which
+    // will be used to check compatibility of the provided converter.
+    const typeVar = TypeVarType.createInstance('__converterInput');
+    typeVar.scopeId = evaluator.getScopeIdForNode(converterNode);
+    const targetFunction = FunctionType.createSynthesizedInstance('');
+    targetFunction.details.declaredReturnType = fieldType;
+    FunctionType.addParameter(targetFunction, {
+        category: ParameterCategory.Simple,
+        name: '__input',
+        type: typeVar,
+        hasDeclaredType: true,
+    });
+    FunctionType.addParameter(targetFunction, {
+        category: ParameterCategory.Simple,
+        name: '',
+        type: UnknownType.create(),
+    });
+
+    if (isFunction(converterType)) {
+        const typeVarContext = new TypeVarContext(typeVar.scopeId);
+        const diagAddendum = new DiagnosticAddendum();
+
+        if (evaluator.assignType(targetFunction, converterType, diagAddendum, typeVarContext)) {
+            const solution = applySolvedTypeVars(typeVar, typeVarContext, { unknownIfNotFound: true });
+            return solution;
+        }
+
+        evaluator.addDiagnostic(
+            AnalyzerNodeInfo.getFileInfo(converterNode).diagnosticRuleSet.reportGeneralTypeIssues,
+            DiagnosticRule.reportGeneralTypeIssues,
+            Localizer.Diagnostic.dataClassConverterFunction().format({
+                argType: evaluator.printType(converterType),
+                fieldType: evaluator.printType(fieldType),
+                fieldName: fieldName,
+            }) + diagAddendum.getString(),
+            converterNode,
+            diagAddendum.getEffectiveTextRange() ?? converterNode
+        );
+    } else {
+        const acceptedTypes: Type[] = [];
+        const diagAddendums: DiagnosticAddendum[] = [];
```

Comment:
The DiagnosticAddendum is designed to handle multiple messages. Rather than accumulating an array of them like this, you should allocate one DiagnosticAddendum before the loop and use it repeatedly. The `assignType` will add all of the messages to it, and getString() will perform any necessary indentation and joining.